### PR TITLE
chore: switch to TypeScript

### PR DESCRIPTION
**Pull request by KingDarBoja**
This is a **work in progress**. References #176

Changelog:
- [x] Switch to TypeScript.
- [x] Move type definitions into `types.ts` file.
- [x] Remove old `index.d.ts` file and set compiler flag to autogenerate those from src.
- [x] Manage to fix the `localeFilter` type definitions ~(Need some help here!)~ casted in the meantime 🧙 .
- [x] Setup Mocha to use ts-node/register.
- [x] Setup ESLint + TS + Prettier.
- [x] Cleanup install (I used yarn so looks like it should be npm instead).
- [x] Converted JSON files into TS files for tree-shaking.

### CODE REVIEW

**Comment by KingDarBoja** [comment]
Would like some review from @danilofuchs as well :D

File: .eslintrc [hunk 1]
```diff
@@ -1,13 +1,24 @@
...
+  "plugins": [
+    "prettier"
+  ],
+  "extends": [
+    "plugin:@typescript-eslint/recommended",
+    "plugin:prettier/recommended",
+    "prettier/@typescript-eslint",
+    "plugin:prettier/recommended"
+
+  ],
```
Comment (danilofuchs) [inline .eslintrc:12]: Oops, we are extending twice
```suggestion
  "extends": [
    "plugin:@typescript-eslint/recommended",
    "plugin:prettier/recommended",
    "prettier/@typescript-eslint"
  ],
```

File: .eslintrc [hunk 2]
```diff
@@ -1,13 +1,24 @@
 {
-  "plugins": ["prettier"],
-  "extends": ["eslint:recommended", "plugin:prettier/recommended"],
+  "parser": "@typescript-eslint/parser",
+  "plugins": [
+    "prettier"
+  ],
```
Comment (danilofuchs) [inline .eslintrc:5]: I believe we have to add `@typescript-eslint` plugin here ([reference](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/README.md#configuration))
```suggestion
  "plugins": [
    "@typescript-eslint",
    "prettier"
  ],
```

File: .eslintrc [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
+  "extends": [
```
Comment (danilofuchs) [inline .eslintrc:6]: Docs also recommend extending `eslint:recommended`

**Review by danilofuchs — COMMENTED** [review]
Awesome initiative and great job!

The Typescript build is failing on a few places, please take a look at them.

**Comment by KingDarBoja** [comment]
> Awesome initiative and great job!
> 
> The Typescript build is failing on a few places, please take a look at them.

I resolved the ESLint config and also made the buil pass by doing some casting because it is not clear some types for `getNames` and `getName` method.

**Comment by michaelwittig** [comment]
this is super hard to review. Have you changed anything expect converting to TS? @danilofuchs suggests that you also fixed a bug here?

**Comment by KingDarBoja** [comment]
> this is super hard to review. Have you changed anything expect converting to TS? @danilofuchs suggests that you also fixed a bug here?

As far as I know, only the linked issues which are about return types being inconsistent with `d.ts` (type definitions).

**Comment by michaelwittig** [comment]
So how is this thing going to be published? So far, all I do is `npm publish`. I don't think that that's enough now?

I don't have much time to invest in this project at the moment and I'm not involved in any frontend work at the moment. Therefore, I don't feel very confident in merging this before I know how to publish it.

**Comment by KingDarBoja** [comment]
Yeah, this is a work in progress as I forget to setup the build steps for compilation and enable multiple browser supports as well. After that, the `npm publish` can be called without issues but wanted to get feedback from other TypeScript users as well :)

**Comment by michaelwittig** [comment]
@KingDarBoja I see.

**Comment by danilofuchs** [comment]
When using Typescript (or any preprocessor like Babel), you have to execute a command to generate output `.js` files before running `npm publish`.

Usually, that would be `npm run build`

**Comment by KingDarBoja** [comment]
I have added the build script with the default `tsconfig.json` options and some extra for emitting the `d.ts` types and load json modules as well. Would be great if someone install it after running the `build` script and play around to spot any issue regarding compatibility.

**Comment by KingDarBoja** [comment]
Almost done with the whole switching as I have installed it locally in some app (using verdaccio) and works almost perfectly, just need to manage the `langs` directory to be included in the public API to avoid importing it like:

```ts
import { getName, registerLocale } from 'i18n-iso-countries';
import { LocaleEN } from 'i18n-iso-countries/dist/langs/en';
registerLocale(LocaleEN);
```

And rather be like:
```ts
import { getName, registerLocale } from 'i18n-iso-countries';
import { LocaleEN } from 'i18n-iso-countries/langs/en';
registerLocale(LocaleEN);
```

**Comment by danilofuchs** [comment]
@KingDarBoja I'm not sure I liked the convertion of locale files from JSON to TS.
JSON is easier for beginners to use and many open PRs are made based on these files.

TS can still check JSON files with the `resolveJsonModules` tsconfig flag, using a cast `localeEN as LocaleData`

**Comment by KingDarBoja** [comment]
> @KingDarBoja I'm not sure I liked the convertion of locale files from JSON to TS.
> JSON is easier for beginners to use and many open PRs are made based on these files.
> 
> TS can still check JSON files with the `resolveJsonModules` tsconfig flag, using a cast `localeEN as LocaleData`

Does this flag let tree-shaking work with json files? My only issue is that I wasn't taking into account the lang files in the build step but since those are json files, I don't think tree-shaking applies for these.

**Comment by KingDarBoja** [comment]
@danilofuchs after trying out some stuff, I made it work as you expected while keeping the `langs` directory files as `.json`:

```ts
import { getName, registerLocale } from 'i18n-iso-countries';
import { COUNTRY_CODES } from 'i18n-iso-countries/codes';
import { localeEN } from 'i18n-iso-countries/bundled';
```

Kinda based on the #188 but without rollup, while relying on some workarounds due to TS lack of support for subpath exports from NodeJS v12+.

**package.json**

```json
{
  "typings": "dist/index.d.ts",
  "main": "dist/entry-node.js",
  "browser": {
    "./dist/entry-node.js": "./dist/index.js"
  },
  "exports": {
    ".": {
      "import": "./dist/index.js"
    },
    "./codes": {
      "import": "./dist/codes.js"
    },
    "./bundled": {
      "import": "./dist/entry-node.js"
    }
  },
  "typesVersions": {
    "*": {
      "codes": ["dist/codes"],
      "bundled": ["dist/entry-node"]
    }
  },
}
```

**Comment by cwqt** [comment]
Any more progress / work arounds for using a single locale?, especially with angular where it builds an entire app per locale, having every locale in the bundle is a bit of a bloat

**Comment by KingDarBoja** [comment]
> Any more progress / work arounds for using a single locale?, especially with angular where it builds an entire app per locale, having every locale in the bundle is a bit of a bloat

While waiting for this PR, I am making a flag pack library that relies on this package but to enable tree-shaking of unused lang files, I have copied the source code into its own library (not released yet): https://github.com/KingDarBoja/ngx-nations

Probably gonna call it `@nation/i18n` as it is not angular specific whereas `@nation/ngx-core` will be ofc Angular support 👯‍♂️

**Comment by Yberion** [comment]
Any news regarding this issue ?

**Comment by bschnabel** [comment]
any plans to merge this into master at some point?

**Comment by KingDarBoja** [comment]
> @bschnabel I don't think that the work here is done. More like a WIP. Correct me if I'm wrong @KingDarBoja ?

Almost 2 years have passed since last update so I can't remember if there was something missing. Probably not, maybe some conflict updates there and there but I am no longer interested in continuing with this PR (no free time for it).

Hope someone else completes this one if the TypeScript conversion is still wanted 😄

**Comment by michaelwittig** [comment]
@KingDarBoja I see. Let me close this PR.

**Pull request closed without merging**